src/com.target.lorri.varlink

```diff
@@ -1,85 +1,104 @@
+    nix_file: string,
+    # The root directory of the project
     project_root: string
 )
 
 type Failure (
+    # The kind of failure:
+    # - io: An I/O failure
+    # - spawn: The build process couldn't be spawned
+    # - exit: The build started but exited with a failure
+    # - output: the build completed, but Lorri wasn't able to interpret the
+    #   output
     kind: (io, spawn, exit, output),
-    msg: ?string,   # only present if kind in (io, spawn)
-    cmd: ?string,   # only present if kind in (spawn, exit)
-    status: ?int,   # only present if kind == exit
-    logs: ?[]string # only present if kind == exit
+    # The absolute path to the shell.nix file for the added project
+    nix_file: string,
+    io: ?IOFail,        # present iff kind == io
+    spawn: ?SpawnFail,  # present iff kind == spawn
+    exit: ?ExitFail,    # present iff kind == exit
+    output: ?OutputFail # present iff kind == output
 )
 
-# WatchServices establishes a stream with the daemon. Initially, the daemon
-# evaluates the given services definition to an array of Command objects and
-# sends a reply for each of them. After this initial evaluation, the daemon
-# watches the services definition and its dependencies for changes,
-# re-evaluates it as appropriate and sends a reply for each Command again.
-#
-# This is a streaming RPC. The daemon only accepts client calls with the "more"
-# property set - see https://varlink.org/Method-Call.
-# TODO: Implement WatchServices
-#method WatchServices(services_nix: ServicesNix) -> (service: Service)
-
-# ServicesNix describes the Nix expression which evaluates to a list of
-# services.
-type ServicesNix (
-  # The absolute path of a Nix file specifying the services to be run. This Nix
-  # file must evaluate to a JSON document of type []Command, that is, an array
-  # of objects whose properties are described by the Command type.
-  path: string
+# Describes a build failure related to opening files, usually the shell.nix file
+type IOFail (
+    # A message describing the failure
+    message: string
 )
 
-# Service describes an individual service to be run.
-type Service (
-  # The user-friendly name of the service. This is used for identification
-  # purposes too: only a single instance of a service with a particular name is
-  # run at any one time.
-  name: string,
-
-  # How to run the service.
-  command: Command
+# Describes a failure to launch the build process
+type SpawnFail (
+    # A message describing the failure
+    message: string,
+    # The command Lorri attempted to execute
+    command: string
 )
 
-# Command describes how to run a terminal application.
-type Command (
-  # The path of the command binary.
-  program: string,
+# Describes a failed build process
+type ExitFail (
+    # The command executed by Lorri
+    command: string,
+    # The Unix exit status of the command, if available
+    status: ?int,
+    # Log output of the command
+    logs: []string
```

Comment:
We should probably distinguish between stdout and stderr here, right? Either two lists, or one list with entries that have a tag whether they are stdout or stderr, which would keep their ordering.

Reply:
It’s only stderr, I changed the docs to make that clear.